plugin/decision: check if event is too large after compression

sspaink opened the pull request:
### Why the changes in this PR are needed?

resolve: https://github.com/open-policy-agent/opa/issues/7526

### What are the changes in this PR?

This PR changes what happens when an event is written to the chunk encoder when calling `Write` and `WriteBytes`. Originally the incoming event _uncompressed_ size was compared to the _compressed_ limit causing the issue. To fix this, the logic has changed to rely on the adaptive uncompressed limit to prevent large events from sneaking into a chunk. In case the uncompressed limit is wrong, the events are decoded and written recursively into a chunk. The base case is that the incoming event is the first event being written into a chunk. This is when the event is compressed and the ND cache or the entire event can be dropped, the benefit is that in case the event is too big even after compression only a single event had to be compressed multiple times.

Moving the logic when to drop the ND cache into the encoder also has the benefit that the `size` and `event` buffer can reuse the logic.

The variable `soft limit` has also been renamed to `uncompressed limit` throughout the code and documentation to help clarify what it is meant to represent.

### Notes to assist PR review:

Repeating the reproduction steps outlined in https://github.com/open-policy-agent/opa/issues/7526, but using a build with the changes in this PR no error is logged.

johanfylling commented on v1/plugins/logs/plugin_test.go:
```diff
@@ -1670,53 +1683,203 @@ func TestPluginRateLimitDropCountStatus(t *testing.T) {
+			expectedDroppedNDCacheEvents: 50,
+			// the adaptive uncompressed limit increases until it stabilizes
+			expectedChunksLengths:            []int{197, 197, 197, 197, 214, 214, 214, 214, 214},
+			expectedEventsInChunk:            []int{1, 1, 1, 1, 9, 9, 9, 9, 9}, // 49 in total, plus one in the encoder
```
What am I missing here? Why do we expect this spread of events, and why does a one-event chunk require `197` bytes while a nine-event chunk requires only slightly more (`214`)? Is this an artifact of how the test is setup or of gzip?

johanfylling replied:
Should the uncompressed soft-limit be similarly traced and asserted as it changes?

johanfylling commented on v1/plugins/logs/encoder_test.go:
```diff
@@ -145,12 +152,12 @@ func TestChunkEncoderAdaptive(t *testing.T) {
-	expectedScaleDownEvents := uint64(3)
+	expectedScaleUpEvents := uint64(4)
+	expectedScaleDownEvents := uint64(0)
 	expectedEquiEvents := uint64(0)
```
Should we assert the expected number of chunks too? I'm guessing these aren't exactly the sum of these events.

sspaink replied:
Updated the test to add more cases, it also checks the sum of the events and the expected IDs of the events to make sure nothing is lost.

johanfylling reviewed (commented):
Some thoughts and questions.

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -126,58 +291,67 @@ func (enc *chunkEncoder) Flush() ([][]byte, error) {
-			enc.metrics.Counter(encSoftLimitScaleUpCounterName).Incr()
-		}
+	// 1) Scale Up
+	if enc.buf.Len() < int(float64(enc.limit)*encCompressedLimitThreshold) {
```
🚀 calculating this on creation and reconfiguration now.

sspaink commented:
@johanfylling I've updated the logic to find an event that is too big to instead make use of the recursion that splits chunks when the uncompressed limit grows too large. Now the uncompressed limit is taken in account, and the first event that is written helps adjust the uncompressed limit to a reasonable starting point opposed to growing from the upload size limit.

Also added a new histogram metric to track the number of events in each chunk. Not sure how useful this is for users 🤔 at the moment I am just using it in `TestChunkEncoderAdaptive` to find the maximum.

Thanks!

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -119,101 +319,80 @@ func (enc *chunkEncoder) writeClose() error {
-	var result [][]byte
-	for i := range events {
-		chunk, err := enc.Write(events[i])
+		chunk, err := enc.scaleDown(events)
```
We need to do this because the last event we wrote to the buffer was too much? Would be nice if there was a way to make sure that we're always below the limit in write. I.e. Whenever `Write()` returns, all hoops have been jumped through. This would mean too large expenses?

sspaink replied:
I agree it would be a lot nicer if `Write()` could guarantee the buffer was always below the limit. So the problem is getting the accurate compressed buffer size. You need to either `Close` or `Flush` the buffer to get the accurate size but they both have significant downsides:

* Calling `Flush` degrades the compression ratio, from the [Flush](https://pkg.go.dev/compress/gzip#Writer.Flush) docs it says `In the terminology of the zlib library, Flush is equivalent to Z_SYNC_FLUSH.` and in the [zlib documentation](https://www.zlib.net/manual.html) it says `Flushing may degrade compression`. It definitely does degrade because I tried to add `enc.w.Flush()` after calling `enc.appendEvent(eventBytes)` to see if we can scale down without pushing in a new event.
* Calling `Close` would require copying the buffer and writer each time to remove the closing bracket.

So by allowing the buffer to potentially exceed the limit lets us skip the above costs and deal with it when an event comes in. Then the downside is that when you call `Flush` on the chunkEncoder it has to fix a buffer that exceeded the limit without an incoming event.

BTW I did not know that the gzip Flush call degraded the compression rate before trying it, definitely a hidden quirk that the Go documentation should do better at highlighting.

johanfylling commented on v1/plugins/logs/plugin.go:
```diff
@@ -1027,54 +1053,17 @@ func (p *Plugin) encodeAndBufferEvent(event EventV1) {
-		p.incrMetric(logNDBDropCounterName)
-	}
-
 	p.mtx.Lock()
```
Before, parts of buffering an event, such as json marshalling, happened outside of this global lock. Do we have a sense of how this will affect lock contention and general performance under high load?

sspaink replied:
I think this will improve performance because calling `p.encodeEvent(event)` acquired and released the lock, and if there wasn't an error, it would immediately try to get the lock again. So while it does hold onto the lock longer because of the JSON marshalling, it doesn't have to get the lock twice. Dropping the ND cache outside of the lock was probably faster, but then it also called `p.encodeEvent(event)` again so it could then get/release the lock three times. So this new way would have less contention.

I don't have any numbers to back this up though 😅 but at least that is what it looks like

johanfylling replied:
But we're also doing more work inside of that lock. We might need to run some benchmarks to get a sense of the impact here.

sspaink replied:
Moving the json marshaling outside of this function is a good idea, it would help during the recursive calls when the event has already been marshaled. To avoid breaking changes added a new method called `Encode` that accepts both the event and byte version. Pushed a new commit with this change.

I ran the same `vegeta attack` benchmark [I did in this PR](https://github.com/open-policy-agent/opa/issues/7562) comparing this latest change against v1.5.0 for the size buffer and this change is a little faster. Small difference but at least shows it isn't much worse!

### latest commit

```
Requests      [total, rate, throughput]         150000, 500.00, 500.00
Duration      [total, attack, wait]             5m0s, 5m0s, 558.875µs
Latencies     [min, mean, 50, 90, 95, 99, max]  107.209µs, 740.847µs, 509.324µs, 701.714µs, 810.828µs, 11.616ms, 57.926ms
```

### v1.5.0

```
Requests      [total, rate, throughput]         150000, 500.00, 500.00
Duration      [total, attack, wait]             5m0s, 5m0s, 766.25µs
Latencies     [min, mean, 50, 90, 95, 99, max]  116.917µs, 979.346µs, 649.718µs, 858.021µs, 992.51µs, 17.061ms, 66.73ms
```

johanfylling reviewed (commented):
Some questions.

It's been a while since I looked at this last, so sorry if I'm rehashing old stuff 😅.

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -55,61 +71,245 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+		return nil, err
+	}
+
+	if int64(len(eventBytes)+enc.bytesWritten+1) <= enc.uncompressedLimit {
```
The `+1` is accounting for the closing bracket `]`, if this event is the last one then `enc.bytesWritten` already has the opening bracket. Maybe I am misunderstanding?

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -55,61 +71,245 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+	var result [][]byte
+	for i := range events {
+		// recursive call to make sure the chunk created adheres to the uncompressed size limit
+		chunk, err := enc.Write(events[i])
```
Nice find! That isn't ideal at all. A solution I could come up with is tracking the size of an event when the ND cache had to be dropped with `enc.maxEventSize`. Then if another event gets written that exceeds that size it will immediately try to drop the cache giving it a chance to be added into the same chunk (if the uncompressed size has grown correctly).

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -55,61 +71,245 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+		enc.scaleUp()
+
+		result, err := enc.reset()
+		if err := enc.appendEvent(eventBytes); err != nil {
```
Your understanding is correct! A recursive call to `Write()` makes sense and avoids the worst case scenario. Great catch, I've updated the code for `1) Scale Up` and `3) Equilibrium` to both do a recursive call to avoid burying a big event.

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -55,61 +71,245 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+
+		currentSize := enc.buf.Len()
+
+		// 2. Try to drop the ND cache and see if the incoming event can fit with the current chunk without it (so we can maximize chunk size)
```
Fixed 👍

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -55,61 +71,245 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+		enc.incrMetric(logNDBDropCounterName)
+		// success! the incoming chunk lost the ND cache, but it wasn't dropped entirely
+		// scale up the uncompressed limit using the uncompressed event size as a base
+		enc.uncompressedLimit = int64(len(eventBytes))
```
good catch! I added the check

sspaink commented:
@johanfylling I think I put too much into one pull request, so I decided to split up the changes into separate issues/PRs. 

* This PR contains the changes to check if an event is too large after compression, updated the issue as well with a way to reproduce this and explains the issue better.
* Don't reset adaptive limit between uploads: https://github.com/open-policy-agent/opa/pull/7521
* Set the minimum and maximum boundaries to the upload limit: https://github.com/open-policy-agent/opa/pull/7563

I think this should help make the reviewing a little easier. I also added better documentation describing the specific problem and how to reproduce it in each issue/PR. Sorry for not doing this to begin with, I don't think it affects any of your most recent review comments. Thank you for bearing with me 😄 

I've updated the description in this PR and updated the

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+		}
+
+		// The ND cache has to be dropped, record this size as a known maximum event size
+		if enc.maxEventSize == 0 || int64(len(eventBytes)) < enc.maxEventSize {
```
So `maxEventSize` is a misnomer? This isn't the maximum allowed size, this is some point beyond the max?

johanfylling replied:
Who knows 😄. Let's keep it, and revise in the future if needed.

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
-		return err
+		r, err := enc.Encode(event, eventBytes)
+		if err != nil {
+			return nil, err
```
If the new event is too large, and can't be added to the encoder, we're dropping the `result` here. Don't we know that it contains a chunk by this point, which we're now dropping despite nothing being wrong with the contained events?

sspaink replied:
Good catch, returning `result` is the right thing to do

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+			result = append(result, r...)
+		}
+
+		return result, err
```
Here, `err` is always `nil`, right?

sspaink replied:
yup! fixed

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+
+		r, err := enc.Encode(event, eventBytes)
+		if err != nil {
+			return nil, err
```
Same here: could we be dropping a perfectly good chunk?

johanfylling commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
-		if enc.softLimitScaleDownExponent < enc.softLimitScaleUpExponent {
-			enc.softLimitScaleDownExponent = enc.softLimitScaleUpExponent
+		// recursive call to make sure the chunk created adheres to the uncompressed size limit
+		chunk, err := enc.Encode(events[i], eventBytes)
```
`chunks`?

johanfylling reviewed (commented):
Some additional comments.

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
-	enc.initialize()
-	if buf != nil {
-		return [][]byte{buf.Bytes()}
+func (enc *chunkEncoder) reset() ([][]byte, error) {
```
That does make more sense, updated to return `[]byte`

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,165 +91,376 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+	// Adjust the encoder's uncompressed limit based on the current amount of
+	// data written to the underlying buffer. The uncompressed limit decides when to return a chunk.
+	// The uncompressed limit is modified based on the below algorithm:
+	// 1) Scale Up: If the current chunk size is within 90% of the user-configured limit, exponentially increase
```
updated in the comments and docs

sspaink commented on v1/plugins/logs/encoder.go:
```diff
@@ -71,138 +91,249 @@ func (enc *chunkEncoder) WithMetrics(m metrics.Metrics) *chunkEncoder {
+// Deprecated: use Write(event EventV1) instead
+func (enc *chunkEncoder) WriteBytes(b []byte) ([][]byte, error) {
+	var e EventV1
+	if err := json.Unmarshal(b, &e); err != nil {
```
awesome that is nicer, dropped

sspaink commented on v1/plugins/logs/eventBuffer.go:
```diff
@@ -124,16 +117,17 @@ func (b *eventBuffer) Upload(ctx context.Context) error {
 			}
 
-			result, err = b.enc.WriteBytes(serialized)
+			result, err = b.enc.Encode(*event.EventV1, eventBytes)
```
I do think the `b.upload` lock is necessary in case the upload timer is very short or if the manual trigger is being called aggressively. Moving `b.enc.Reconfigure()` below the lock seems like the best idea. Looking at it again `b.client` and `b.uploadPath` should also go under the lock for the same reason.

johanfylling reviewed (commented):
Ran some perf tests of my own; plerformance looks largely unchanged.

<img width="811" alt="Screenshot 2025-06-19 at 15 39 03" src="https://github.com/user-attachments/assets/6341685c-bf3c-4975-bd1a-65a8275e2b26" />

main: 

```Global metrics:
Requests: 200
Requests total: 100000
Duration: 1m18.516058833s
Max concurrency: 500
Average req/s: 1273.6248034646687
timer_rego_query_eval_ns: Min: 5.375µs, Max: 817.916µs, Mean: 21.718µs, P50: 13.479µs, P75: 17.489µs, P90: 39.924µs, P95: 61.302µs, P99: 161.134µs, P99.9: 809.603µs, P99.99: 817.916µs
duration: Min: 180.792µs, Max: 88.047416ms, Mean: 678.158µs, P50: 385.375µs, P75: 477.729µs, P90: 796.262µs, P95: 1.218756ms, P99: 3.42696ms, P99.9: 87.394911ms, P99.99: 88.047416ms
timer_server_handler_ns: Min: 68.542µs, Max: 11.609167ms, Mean: 231.362µs, P50: 170.854µs, P75: 217.093µs, P90: 328.608µs, P95: 526.806µs, P99: 1.156035ms, P99.9: 11.396883ms, P99.99: 11.609167ms
timer_rego_external_resolve_ns: Min: 41ns, Max: 3.958µs, Mean: 115ns, P50: 84ns, P75: 125ns, P90: 167ns, P95: 208ns, P99: 334ns, P99.9: 3.868µs, P99.99: 3.958µs
timer_rego_query_compile_ns: Min: 10.875µs, Max: 385.125µs, Mean: 30.839µs, P50: 25.187µs, P75: 30.656µs, P90: 41.271µs, P95: 58.918µs, P99: 235.747µs, P99.9: 384.389µs, P99.99: 385.125µs
Peaks:
duration: 93.717667ms
timer_server_handler_ns: 93.573541ms
timer_rego_external_resolve_ns: 343.375µs
timer_rego_query_compile_ns: 7.865875ms
timer_rego_query_eval_ns: 4.087875ms
Peak duration: 93.717667ms
```

PR: 

```
Global metrics:
Requests: 200
Requests total: 100000
Duration: 1m18.222317792s
Max concurrency: 500
Average req/s: 1278.4075289856378
duration: Min: 164.5µs, Max: 63.79725ms, Mean: 698.156µs, P50: 398.854µs, P75: 497.385µs, P90: 905.012µs, P95: 1.423327ms, P99: 4.176799ms, P99.9: 63.694633ms, P99.99: 63.79725ms
timer_rego_query_eval_ns: Min: 5.75µs, Max: 408.417µs, Mean: 20.925µs, P50: 13.833µs, P75: 17.364µs, P90: 39.675µs, P95: 55.189µs, P99: 180.39µs, P99.9: 406.198µs, P99.99: 408.417µs
timer_server_handler_ns: Min: 70.666µs, Max: 33.640792ms, Mean: 247.448µs, P50: 180.375µs, P75: 218.073µs, P90: 293.366µs, P95: 499.251µs, P99: 1.030884ms, P99.9: 32.711279ms, P99.99: 33.640792ms
timer_rego_external_resolve_ns: Min: 41ns, Max: 2.583µs, Mean: 116ns, P50: 125ns, P75: 125ns, P90: 167ns, P95: 167ns, P99: 292ns, P99.9: 2.547µs, P99.99: 2.583µs
timer_rego_query_compile_ns: Min: 11.208µs, Max: 739.959µs, Mean: 32.406µs, P50: 26.145µs, P75: 30.833µs, P90: 39.595µs, P95: 57.502µs, P99: 227.093µs, P99.9: 733.351µs, P99.99: 739.959µs
Peaks:
timer_rego_query_eval_ns: 2.902666ms
timer_server_handler_ns: 91.444459ms
timer_rego_external_resolve_ns: 299.792µs
timer_rego_query_compile_ns: 2.533792ms
duration: 92.222292ms
Peak duration: 92.222292ms
```

Looks like we're nearing the end of this story 🙂. I think there might be just the one thing left to fix.

johanfylling reviewed (approved):
Nice! 😃 

Let's get this baby in! 🎉
